design/ceph/object/swift-and-keystone-integration.md

```diff
@@ -0,0 +1,241 @@
+    apiVersion: 3
+    implicitTenants: swift
+    tokenCacheSize: 1000
+    revocationInterval: 1200
+    serviceUserSecret: rgw-service-user
+```
+
+The certificate to verify the Keystone endpoint can't be explicitly
+configured in Ceph RGW, the system configuration of the pod is used,
+you can add to the system certificate store via the
+`gateway.caBundleRef` setting of the object store CRD.
+
+The credentials for the Keystone service account used by Ceph RGW are
+supplied in a Secret that contains a mapping of OpenStack openrc
+environment variables. Only password authentication to Keystone is
+supported. Example:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: rgw-service-user
+data:
+  OS_PASSWORD: "horse staples battery correct"
+  OS_USERNAME: "ceph-rgw"
+  OS_PROJECT_NAME: "admin"
+  OS_USER_DOMAIN_NAME: "Default"
+  OS_PROJECT_DOMAIN_NAME: "Default"
+  OS_AUTH_TYPE: "password"
+```
+
+This format is chosen because it is a natural and interoperable way
+that keystone credentials are represented and for the supported auth
+type it maps naturally to the Ceph RGW configuration.
+
+The following constraints must be fulfilled by the secret:
+* `OS_AUTH_TYPE` must be `password` or omitted.
+* `OS_USER_DOMAIN_NAME` must equal `OS_PROJECT_DOMAIN_NAME`.
+* All openrc variables not in the example above (e.g. API version and
+  Keystone endpoint) are ignored.
+
+### Swift integration
+
+Swift is configured in the `gateway:` section of the Object Store CRD.
+
+The currently unused `type:` argument may now take the values `swift`
+or `s3`.
+
+The following new settings are available to properly configure Swift:
+```yaml
+gateway:
+  type: swift
+  swiftAccountInUrl: true
+  swiftUrlPrefix: /swifter
+  swiftVersioningEnabled: false
+```
+
+The access to the Swift API is granted by creating a subuser of an RGW
+user. While commonly the access is granted via projects
+mapped from Keystone, explicit creation of subusers is supported by
+extending the `cephobjectstoreuser` resource with a new section
+`spec.subUsers`:
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: CephObjectStoreUser
+metadata:
+  name: my-user
+  namespace: rook-ceph
+spec:
+  store: my-store
+  displayName: my-display-name
+  quotas:
+    maxBuckets: 100
+    maxSize: 10G
+    maxObjects: 10000
+  capabilities:
+    user: "*"
+    bucket: "*"
+  subUsers:
+  - name: swift
+    access: full
+```
+
+The subusers are not mapped to a separate CR for the
+following reasons:
+
+* The full subuser names are prefixed with the username like
+  `my-user:subusername`, so being unique within the CR guarantees
+  global uniqueness.
+
+  Unlike `radosgw-admin` the subuser name in the CRD must be provided
+  *without the prefix* (radosgw-admin allows both, to omit or include
+  the prefix).
+
+* The subuser abstraction is very simple – only a name and an access
+  level can be configured, so a separate resource would not be
+  appropriate complexity wise.
```

Comment:
Link to Ceph docs?

Reply:
There is a link to the relevant docs in the vicinity now.